daedalus/src/redemption_key_validation.js

```diff
@@ -0,0 +1,208 @@
+//  > Buffer.from("====", "base64").toString("utf-8")
+//  ''
+//  > Buffer.from("aa==", "base64").toString("utf-8")
+//  'i'
+//  > Buffer.from("aa=", "base64").toString("utf-8")
+//  'i'
+//
+// And that's the reason this funciton exists: to extend Node's base64 decode with validity check.
+function decodeNode (str) {
+    var result;
+
+    // Check that the input string is a valid Base64-encoded string as Node.js
+    // decided that it would be a good idea to NOT throw on invalid input strings
+    // but return an empty buffer instead which cannot be distinguished from the
+    // empty string case.
+    var reEmptyString = "^$";
+    var leadingQuanta = "^([A-Za-z0-9+/]{4})*";
+    var finalQuantum =
+        "([A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}(?:=)?|[A-Za-z0-9+/]{2}(?:=){0,2})$";
+    var reValidBase64 =
+        new RegExp([reEmptyString, "|", leadingQuanta, finalQuantum].join(""));
+
+    if (!reValidBase64.test(str)) { throw new Error("Invalid input string");}
+    result = Buffer.from(str, "base64").toString("utf-8");
+
+    return result;
+};
+
+// Convert base64url into base64
+function toRfc4648(str) {
+    return str.replace(new RegExp('_', 'g'), '/').replace(new RegExp('-', 'g'), '+');
+}
+
+// Convert String into Uint8Array buffer
+function unsafeStringToUint8ArrayOfCharCodes(str) {
+    var buf = new ArrayBuffer(str.length);
+    var bufView = new Uint8Array(buf);
+    for (var i=0, strLen=str.length; i<strLen; i++) {
+        bufView[i] = str.codePointAt(i);
+    }
+    return bufView;
+}
+
+// The atob() method decodes a base-64 encoded string.
+// It lives within browser implementation (non existent in nodejs)
+// I believe it exists in Electron as its chrome box.
+var atobIsDefined = typeof atob === "function";
+
+// Decode buffer into utf8 string
+function decodeUtf8(buf) {
+    var decoder = new TextDecoder('utf-8');
+    decoder.decode(buf);
+}
+
+// Implements: "Valid redemption key should be base64 and base64url decodable, it should end with '=' and it should be 44 chars long."
+function isValidRedemptionKey(code) {
+    var res = true;
+    try {
+        // Check if atob is defined (if we are in browser - it should be as we are in Electron)
+        // Note that with Electron we will probably run the first branch (so, atob is defined) but just in case we add fallback - as there are some mentiones that it was not working sometimes https://github.com/electron/electron/issues/8529
+        if(atobIsDefined) {
+            // If we are in browser then:
+            //  * convert input data from base64url into base64 - as vending was done in a messy way with both base64 and base64 url
+            //  * decode base64 with atob - note that atob usually does base64 validity check under the hood
+            //  * convert string into Uint8Array with unsafeStringToUint8ArrayOrCharCodes
+            //  * decode Uint8Array into a utf-8 string
+            decodeUtf8(unsafeStringToUint8ArrayOrCharCodes(atob(toRfc4648(code))));
```

Comment:
In general, do we need conversion to `utf-8`? Why `ascii` is not enough?

Reply:
In our case it is, true. I was just doing the exact translation in the first pass.